nemo_text_processing/g2p/data/heteronym_classification_data.py

```diff
@@ -65,98 +67,136 @@ def __init__(
+            target_word_ids = [self.LOSS_PAD_TOKEN]  # -100 to pad plain tokens
+        else:
+            target_word_ids = None  # for inference when labels are not available
+
+        heteronym_span_idx = 0
+        # split sentence by space and keep track of word boundaries
+        # we assume heteronym is a standalone word
+        matches = [(m.group(0), (m.start(), m.end() - 1)) for m in re.finditer(r'\S+', sentence)]
+        for match in matches:
+            word, word_start_end = match
+            # check if the start of the next heteronym span is within the word indices
+            if (
+                heteronym_span_idx < len(start_end)
+                and word_start_end[0] <= start_end[heteronym_span_idx][0] < word_start_end[1]
+            ):
+                heteronym_start_end = start_end[heteronym_span_idx]
+                prefix = ""
+                prefix_ids = []
+                # for cases when word also includes punctuation marks at the beginning or a prefix,
+                # e.g. "diffuse" vs. diffuse vs. pre-diffuse for heteronym {diffuse}
+                if word_start_end[0] < heteronym_start_end[0]:
+                    prefix = sentence[word_start_end[0] : heteronym_start_end[0]]
+                    prefix_ids = self.tokenizer.text_to_ids(prefix)
+                    subtokens_mask.extend([self.PAD_TOKEN] * len(prefix_ids))
+
+                word = word[word.index(prefix) + len(prefix) :]
+                word_input_ids = self.tokenizer.text_to_ids(word)
+                input_ids.extend(prefix_ids + word_input_ids)
+
+                subtokens_mask.extend([1] + [self.PAD_TOKEN] * (len(word_input_ids) - 1))
+
+                if self.with_labels:
+                    cur_target_word_id = self.wordid_to_idx[word_ids[heteronym_span_idx]]
+                    target_word_ids.extend(
+                        [self.LOSS_PAD_TOKEN] * len(prefix_ids)
+                        + [cur_target_word_id]
+                        + [self.LOSS_PAD_TOKEN] * (len(word_input_ids) - 1)
+                    )
+
+                heteronym = sentence.lower()[heteronym_start_end[0] : heteronym_start_end[1]]
+                if heteronym not in self.wiki_homograph_dict:
+                    logging.debug(f"{heteronym} is not supported. Skipping example.")
+                    return None
+
+                heteronym_span_idx += 1
+            else:
+                ids = self.tokenizer.text_to_ids(word)
+                input_ids.extend(ids)
+                subtokens_mask.extend([self.PAD_TOKEN] * len(ids))
+                if self.with_labels:
+                    target_word_ids.extend([self.LOSS_PAD_TOKEN] * len(ids))
+
+        if heteronym_span_idx < len(start_end):
+            logging.info("Not all heteronym spans were processed. Skipping example.")
+            return None
+
+        # add eos token
+        input_ids.append(self.tokenizer.eos_id)
```

Comment:
same as above, how to guarantee that eos_id exists as a member variable of the class TokenizerSpec. What do we do if it doesn't exisit?

Reply:
added a check